### ISSUE TITLE
TypeError: Cannot set property WebSocket of #<Object> which has only a getter / Gradio App not working in NextJS

### ISSUE THREAD

**Issue by TommeZ** [gradio-app/gradio#5583]
### Describe the bug

I can't find anything that resolves my specific issue.

I'm trying to make a fetch request to my HuggingFace space via the Gradio app on my NextJS project, but I'm getting the following error: 

> TypeError: Cannot set property WebSocket of #<Object> which has only a getter
    at eval (webpack-internal:///(rsc)/./node_modules/@gradio/client/dist/index.js:282:34)
- error node_modules\@gradio\client\dist\index.js (271:24) @ eval
- error unhandledRejection: Error [TypeError]: Cannot set property WebSocket of #<Object> which has only a getter
    at eval (webpack-internal:///(rsc)/./node_modules/@gradio/client/dist/index.js:282:34) {
  digest: undefined
}

Does anyone know if there's a way to resolve this, or know what's going on here that I can fix?

### Have you searched existing issues?  🔎

- [X] I have searched and found no existing issues

### Reproduction

For private reasons, I have removed the names of the actual response picture and Hugging Face space.

`import { client } from "@gradio/client";

const HF_ACCESS_TOKEN = process.env.HF_ACCESS_TOKEN

export default async function getDetections() {

    const response = await fetch("jpg");
    const exampleImage = await response.blob();

    const app = await client("HF_Name/MLDetector", { hf_token: HF_ACCESS_TOKEN });

    const result = await app.predict("/predict", [
        exampleImage, 	// blob in 'img' Image component
    ]);

}`

### Screenshot

![Capture](https://github.com/gradio-app/gradio/assets/29313293/56ff3e33-189a-4c3e-9bfb-fc6b8e6b163f)

### System Info

```shell
Gradio Environment Information:
Operating System:  Windows


gradio version:  3.38.0   

gradio Dependencies:      
  aiofiles: 23.1.0        
  aiohttp: 3.8.5
  altair: 5.0.1
  fastapi: 0.100.0        
  ffmpy: 0.3.1
  gradio-client: 0.2.10
  httpx: 0.24.1
  huggingface-hub: 0.16.4
  jinja2: 3.0.2
  markdown-it-py: 2.2.0
  markupsafe: 2.0.1
  matplotlib: 3.7.1
  mdit-py-plugins: 0.3.3
  numpy: 1.24.3
  orjson: 3.9.2
  packaging: 23.1
  pandas: 1.3.4
  pillow: 9.5.0
  pydantic: 1.10.11
  pydub: 0.25.1
  python-multipart: 0.0.6
  pyyaml: 6.0
  requests: 2.26.0
  semantic-version: 2.10.0
  typing-extensions: 4.7.1
  uvicorn: 0.23.1
  websockets: 11.0.3


gradio_client version:  0.2.10

gradio_client Dependencies:
  fsspec: 2023.6.0
  httpx: 0.24.1
  huggingface-hub: 0.16.4
  packaging: 23.1
  requests: 2.26.0
  typing-extensions: 4.7.1
  websockets: 11.0.3
```

### Severity

Blocking usage of gradio

**Comment by pngwn** [comment]
Since this seems to to be an environment or framework specific issue, we will need a more complete reproduction. Could you put together a github repo with a default next app showing this problem? Preferably one that connects to a public HF space so we can run it. 

Thanks.

**Comment by abidlabs** [comment]
Will close for now -- can reopen once we have a repro.

**Comment by piscopancer** [comment]
not fixed?

**Comment by TommeZ** [comment]
I'm currently working on a way to replicate the issue, by creating a new repo that doesn't use my company's code. Will share once complete. But it appears so far that fetching data from Hugging Face via Gradio generates this error when run inside a Next.js project.

**Comment by pngwn** [comment]
Do you have a dump of your system info? Is next running locally or deployed to vercel? If deployed, do you know if it is using the worker runtime or the lambda runtime?

**Comment by JimmyLv** [comment]
I also met this issue. running in my local env.
```shell
sw_vers
ProductName:		macOS
ProductVersion:		13.2.1

node -v
v18.14.2
```

**Comment by TommeZ** [comment]
I've replicated the issue at the following repo:

https://github.com/TommeZ/gradio-fetch

getTest().js makes a request to Hugging Face which returns the mentioned error, whereas it should be returning a string in the terminal.

**Comment by Jay2theWhy** [comment]
Found a similar issue here: https://github.com/vercel/next.js/issues/41866 with a really simple fix. Created a PR for it: https://github.com/gradio-app/gradio/pull/5840

**Comment by JimmyLv** [comment]
@Jay2theWhy do you mean the fix: `replaced axios with cross-fetch`?

**Comment by Jay2theWhy** [comment]
@JimmyLv I mean this PR that was merged for the fix: https://github.com/spacebudz/lucid/pull/108. From my testing, this resolved my conflict with my gradio app and next.js. I also ran the unit tests on this change and all passed.

**Comment by JimmyLv** [comment]
cool, I just saw your PR https://github.com/gradio-app/gradio/pull/5840 good job! 👍

**Comment by dylan0356** [comment]
Still having this exact issue with NextJS implemented with hugging face.

**Comment by TommeZ** [comment]
I've just updated to the latest gradio/client via npm, and I'm receiving the same error still?

**Comment by abidlabs** [comment]
We haven't released a new version of the client just yet, but will be soon (in the next day or so)

**Comment by Jay2theWhy** [comment]
@TommeZ You can follow this PR for when the next version is released: https://github.com/gradio-app/gradio/pull/5831
Should be resolved for v0.5.2

**Comment by TommeZ** [comment]
Thanks for your support gentlemen.

**Comment by JimmyLv** [comment]
Just upgraded to v0.5.2 but still running into this problem, I'm pretty sure `global.Websocket` is `undefined` but can't assign a value. 


```shell
========global.Websocket======== undefined undefined

TypeError: Cannot set property WebSocket of #<Object> which has only a getter
```

Seems like the error might be due to trying to set a property on a frozen object or a non-extensible `global` object.

So my solution was to define `let MyWebSocket` with a new variable and the problem was solved!

```ts
let MyWebSocket
function api_factory(fetch_implementation, WebSocket_factory) {
   ...
     async function client2(app_reference, options = { normalise_files: true }) {
       if ((typeof window === "undefined" || !("WebSocket" in window)) && !globalThis.Websocket) {
          const ws = await import("./wrapper-6f348d45.js");
          NodeBlob = (await import("node:buffer")).Blob;
          MyWebSocket = ws.WebSocket;
        }
      }
}

const { post_data, upload_files, client, handle_blob } = api_factory(
  fetch,
  (...args) => new MyWebSocket(...args)
);
```

At the same time I realized that another solution is to customize a `client` using `api_factory` but not using the passed in WebSocket_factory in the constructor method, hahaha I don't know why.

**Comment by pngwn** [comment]
If it is undefined but can't be assigned then the client won't work without some other changes. I'll take a closer look at this as soon as I can.

**Comment by odysseaspapadimas** [comment]
Any updates or workarounds on this?

**Comment by TommeZ** [comment]
Any update guys?

**Comment by cantonalex** [comment]
I also have the issue. Super annoying.

**Comment by cantonalex** [comment]
@pngwn

**Comment by pngwn** [comment]
Closed by #8118!
